Add CLUSTER FLUSHSLOT command

wuranxx opened the pull request:
Add cluster flushslot command. Closes #1133.

Syntax:

    CLUSTER FLUSHSLOT <slot> [ASYNC|SYNC]

## Background

We want to implement the ability to SYNC or ASYNC free all data from a single slot. This is useful in two cases:

1. When a primary loses ownership of a slot, it will temporarily freeze while it synchronously frees the slot.
2. When doing resharding, you may prefer to forcibly delete all the data from a slot and assign it to another node.

## Internal changes

Modify the function signature of `delKeysInSlot`, adding parameters `lazy`, `propagate_del` and `send_del_event` to control the behavior.

wuranxx committed:
add cluster flushslot command.

Signed-off-by: wuranxx <wuranxx@foxmail.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -6842,6 +6842,23 @@ int clusterCommandSpecial(client *c) {
+    } else if (!strcasecmp(c->argv[1]->ptr, "flushslot") && (c->argc == 3 || c->argc == 4)) {
+        /* CLUSTER FLUSHSLOT <slot> [ASYNC|SYNC] */
+        int slot;
+        int lazy = 1;
```
Let's use `server.lazyfree_lazy_user_flush` by default? That's the default for FLUSHALL when the caller doesn't specify sync or async.

wuranxx replied:
fixed.

wuranxx committed:
Add TCL tests and optimize code structure.

Signed-off-by: wuranxx <wuranxx@foxmail.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx commented:
Please help review this PR. Thank you. @hwware @madolson @PingXie @zuiderkwast

wuranxx committed:
Fix compile warning.

Signed-off-by: wuranxx <wuranxx@foxmail.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

hwware commented:
There are too much information in issue https://github.com/valkey-io/valkey/issues/1133. To save the reviewer time, could you pls update the top comment to describe the background why we want to this new command? You could reference PR https://github.com/valkey-io/valkey/pull/1392.   

Thanks a lot

PingXie commented on src/cluster_legacy.c:
```diff
@@ -7041,3 +7049,21 @@ int clusterDecodeOpenSlotsAuxField(int rdbflags, sds s) {
+            addReplyErrorObject(c, shared.syntaxerr);
+            return;
+        }
+    }
```
we should fail the `c->argc > 4` case I think

wuranxx replied:
In the call site of `clusterCommandFlushslot`, I have already performed a check on the number of parameters. Is it still necessary to validate the number of parameters again within the function?

```c
 else if (!strcasecmp(c->argv[1]->ptr, "flushslot") && (c->argc == 3 || c->argc == 4)) {
        /* CLUSTER FLUSHSLOT <slot> [ASYNC|SYNC] */
        clusterCommandFlushslot(c);
    }
```

zuiderkwast replied:
It's a bit weird that the arity check is split into multiple places. But I don't mind too much because it's like that for many commands already.

But I think this should be in cluster.c instead of cluster_legacy.c. Cluster_legacy is about the cluster bus, and we would replace this file if we replace the cluster bus for cluster V2. At least, that's the idea. This command doesn't use the cluster bus at all, so I think it should be in cluster.c and be called directly from `clusterCommand()`.

PingXie commented on tests/unit/cluster/cluster-flush-slot.tcl:
```diff
@@ -0,0 +1,34 @@
+
+        # flush slot key
+        R 0 CLUSTER FLUSHSLOT $key_slot ASYNC
+        assert_equal [R 0 CLUSTER COUNTKEYSINSLOT $key_slot] 0
```
we should add a wait (`wait_for_condition`). This could be a source of test flakiness.

wuranxx replied:
this command execute failed when I rebased to the latest unstable https://github.com/valkey-io/valkey/commit/6df376d68a97e9c0da4549f57db96742b5482202, I try to confirm the cause.

PingXie reviewed (commented):
LGTM overall. Thanks @wuranxx!

wuranxx commented:
Before I rebased to the latest  unstable (old commit is 32f7541fe34e5e0520a5917d09661756d330bd11) , the Tcl tests and manual command tests ran fine locally. 

After I rebased to the latest unstable 6df376d68a97e9c0da4549f57db96742b5482202, the command failed when executed locally. 

It is possible that changes in the code during this period caused the inconsistent behavior. Until I confirm the cause, I have marked this PR as WIP (Work In Progress).

wuranxx committed:
Fix delKeysInSlot does not consider command invoke.

Signed-off-by: wuranxx <wuranxx@foxmail.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx commented:
> Before I rebased to the latest unstable (old commit is [32f7541](https://github.com/valkey-io/valkey/commit/32f7541fe34e5e0520a5917d09661756d330bd11)) , the Tcl tests and manual command tests ran fine locally.
> 
> After I rebased to the latest unstable [6df376d](https://github.com/valkey-io/valkey/commit/6df376d68a97e9c0da4549f57db96742b5482202), the command failed when executed locally.
> 
> It is possible that changes in the code during this period caused the inconsistent behavior. Until I confirm the cause, I have marked this PR as WIP (Work In Progress).

Fixed. For details, see #1399.

zuiderkwast commented:
@enjoy-binbin @murphyjacob4 Do you want this command for atomic slot migration, for example when a slot migration is cancelled, the importing node can send CLUSTER FLUSHSLOT to its replicas to clean up the slot?

murphyjacob4 commented:
I think we will have a protocol for cancelling the slot migration, and this will trigger the drop for all slot ranges being migrated locally. But we should be able to use the async `delKeysInSlot` functionality to implement this

zuiderkwast commented:
> I think we will have a protocol for cancelling the slot migration, and this will trigger the drop for all slot ranges being migrated locally. But we should be able to use the async `delKeysInSlot` functionality to implement this

@murphyjacob4 The replicas are not involved in the slot migration protocol IIUC. So when a primary cleans up a cancelled slot import, do you just replicate a lot of DEL commands to the replicas? Replicating a single FLUSHSLOT may save some work and bandwidth...

murphyjacob4 commented:
> do you just replicate a lot of DEL commands to the replicas? Replicating a single FLUSHSLOT may save some work and bandwidth...

Yeah that's a good idea. Currently, this is what `delKeysInSlot` does, and it could result in a lot of bandwidth for big slots. Disabling the replication of the individual keys in `delKeysInSlot` and replacing with a single `FLUSHSLOT` makes sense to me. This path is also executed in the dirty slot case, so it won't just be an improvement for atomic slot migration (although I anticipate it will be executed more frequently there).

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -6353,8 +6355,13 @@ unsigned int delKeysInSlot(unsigned int hashslot) {
+        propagateDeletion(&server.db[0], key, lazy);
         signalModifiedKey(NULL, &server.db[0], key);
         /* The keys are not actually logically deleted from the database, just moved to another node.
          * The modules needs to know that these keys are no longer available locally, so just send the
```
This code comment about keyspace notifications:

```
        /* The keys are not actually logically deleted from the database, just moved to another node.
         * The modules needs to know that these keys are no longer available locally, so just send the
         * keyspace notification to the modules, but not to clients. */
```

`delKeysInSlot` is used for slot migrations, which are considered not deleting the keys but just migrating them to another node.

This is not necessarily the case for CLUSTER FLUSHSLOT. I think we should treat CLUSTER FLUSHSLOT more like FLUSHDB. For FLUSHDB, what are the keyspace notifications sent? A lot of DEL events or one FLUSHDB event?

Maybe we shouldn't use `delKeysInSlot` here, or use some extra argument or flag to indicate if it's a migration or if we're just deleting the keys.

wuranxx replied:
1.  For FLUSHDB, what are the keyspace notifications sent? A lot of DEL events or one FLUSHDB event?
    - In my test, flushdb not trigger any event. 

```
192.168.0.1:2774> set k2 1
OK
192.168.0.1:2774> flushdb
OK
```

```
192.168.0.1:2774> PSUBSCRIBE __key*@0__:*
Reading messages... (press Ctrl-C to quit)
1) "psubscribe"
2) "__key*@0__:*"
3) (integer) 1
1) "pmessage"
2) "__key*@0__:*"
3) "__keyspace@0__:k2"
4) "set"
1) "pmessage"
2) "__key*@0__:*"
3) "__keyevent@0__:set"
4) "k2"
```

2.  Maybe we shouldn't use delKeysInSlot here, or use some extra argument or flag to indicate if it's a migration or if we're just deleting the keys.
    - yes, we should add some flag to mark is migration or del.


If we want to like flushdb, we should write a new method to implement flushslot logic.

zuiderkwast replied:
> In my test, flushdb not trigger any event.

That's surprising, but OK. I guess flushslot should behave the same as flushdb. I'm open to other suggestions though. If this was a bad design for flushdb, then we can do something else for flushslot.

> If we want to like flushdb, we should write a new method to implement flushslot logic.

Let's to it?

wuranxx replied:
Okay, I'll try to rewrite the flushslot as per the flushdb implementation.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -6353,8 +6355,13 @@ unsigned int delKeysInSlot(unsigned int hashslot) {
+        if (lazy) {
+            dbAsyncDelete(&server.db[0], key);
+        } else {
+            dbSyncDelete(&server.db[0], key);
```
Here, the deleted keys are replicated just by multiple DEL commands. For comparison, FLUSHDB is replicated as a single FLUSHDB command, which is much less traffic than many DEL commands. If we want to replicate FLUSHSLOT as a single FLUSHSLOT command, then we should probably check that the replicas supports this though. We already have the replica's version stored in `c->repl_data->replica_version` (REPLCONF VERSION is sent by the replica since 8.0) so we could check that it's >= 8.1 for all connected replicas. It's also fine to skip this. We can do it later as a separate optimization.

zuiderkwast replied:
Actually I think we don't have to worry about old replicas. This is a new feature and users should use new features only after all nodes have been upgraded.

murphyjacob4 replied:
For atomic slot migration - we will need send some FLUSHSLOT commands implicitly:

1. If an import doesn't succeed (is failed or cancelled) we need to flush the contents of the staged slots to replicas
2. If there is an import to a target primary and FLUSHDB is run on that target, the target will need to break this out into FLUSHSLOT commands to avoid the importing slots being flushed
3. If there is an export happening from some source primary and FLUSHDB is run, we need to break this out into FLUSHSLOT commands (so that the target only flushes the impacted importing slots)

Plus - in the case that a dirty slot is found - we currently delete it by sending individual key deletions. It would make sense to create the FLUSHSLOT command implicitly to improve replication traffic.

Due to the cases of implicit FLUSHSLOT propagation - I would suggest that the "delKeysInSlot" helper method include logic for deciding about if we should propagate as individual deletions or as FLUSHSLOT based on the capabilities of the replicas. I'm sure for many customers, they may be running in a cross-version deployment for some amount of time while they do rolling upgrades

zuiderkwast replied:
@murphyjacob4

> we will need send some FLUSHSLOT commands implicitly

Yes, so do you need any changes here? You can call this function and then you explicitly replicate a FLUSHSLOT command. Or do you want this function to do it?

I think you may want to skip keyspace events and skip replication using DEL commands. With the `is_cmd` argument you can't disable both. See other comment.

> include logic for deciding about if we should propagate as individual deletions or as FLUSHSLOT based on the capabilities of the replicas. I'm sure for many customers, they may be running in a cross-version deployment for some amount of time while they do rolling upgrades

I agree that we should support cross-version deployments, but in this case I'm not worried and I don't think we need to decide based on replica capabilities.

We require that all nodes are upgraded before you start using new features. For example, an old replica wouldn't understand a new datatype, or information like hash field expiration, etc. The FLUSHSLOT command and the atomic slot migration are both new features. If a node has ASM, it also has FLUSHSLOT. If you have an 8.1 replica in your cluster, you shouldn't start using these features just yet. Makes sense?

zuiderkwast reviewed (commented):
Hi @wuranxx and sorry for the delay.

It looks strait-forward, but there are some details we should think about a bit more. I'm not actually sure what's best to do in all cases, but see my comments below.

wuranxx committed:
1. Move clusterCommandFlushslot to cluster.c
2. add command scene to delKeysInSlot. In command sence, delKeysInSlot will not propagate del command, and notify del event.

Signed-off-by: wuranxx <wuranxx@foxmail.com>

wuranxx commented:
I add command scene to delKeysInSlot. In command sence, delKeysInSlot will not propagate del command,  and notify del event.

flushslot now just propagate `flushslot` command to aof and replica.

please help review this PR, thank you! @zuiderkwast @PingXie

PingXie commented on src/cluster_legacy.c:
```diff
@@ -6449,13 +6450,24 @@ unsigned int delKeysInSlot(unsigned int hashslot) {
-         * keyspace notification to the modules, but not to clients. */
-        moduleNotifyKeyspaceEvent(NOTIFY_GENERIC, "del", key, server.db[0].id);
+        if (is_cmd) {
+            /* In cluster flushslot scene, the keys are actually deleted. */
```
```suggestion
            /* In the `cluster flushslot` scenario, the keys are actually deleted so notify everyone. */
```

PingXie commented on src/cluster_legacy.c:
```diff
@@ -6449,13 +6450,24 @@ unsigned int delKeysInSlot(unsigned int hashslot) {
+            /* The keys are not actually logically deleted from the database, just moved to another node.
+            * The modules needs to know that these keys are no longer available locally, so just send the
+            * keyspace notification to the modules, but not to clients. */
+            /* In cluster flushslot scene, the is actually deleted, fire del event. */
```
```suggestion
```

PingXie commented on src/commands/cluster-flushslot.json:
```diff
@@ -0,0 +1,46 @@
+        "summary": "Remove all keys from the target slot.",
+        "complexity": "O(N) where N is the number of keys in the target slot",
+        "group": "cluster",
+        "since": "8.1.0",
```
```suggestion
        "since": "9.0.0",
```

PingXie commented on src/commands/cluster-flushslot.json:
```diff
@@ -0,0 +1,46 @@
+        "arity": -3,
+        "container": "CLUSTER",
+        "function": "clusterCommand",
+        "command_flags": [
+            "WRITE"
+        ],
+        "acl_categories": [
+            "KEYSPACE",
+            "DANGEROUS"
+        ],
```
I think we should block this command on async loading. also, `ADMIN` is a more appropriate flag, which gets `DANGEROUS` automatically.
```suggestion
        "command_flags": [
            "WRITE",
            "NO_ASYNC_LOADING",
            "ADMIN"
        ],
        "acl_categories": [
            "KEYSPACE"
        ],
```

PingXie commented on src/commands/cluster-flushslot.json:
```diff
@@ -0,0 +1,46 @@
+                        "name": "async",
+                        "type": "pure-token",
+                        "token": "ASYNC",
+                        "since": "8.1.0"
```
```suggestion
                        "since": "9.0.0"
```

PingXie commented on src/commands/cluster-flushslot.json:
```diff
@@ -0,0 +1,46 @@
+                        "name": "sync",
+                        "type": "pure-token",
+                        "token": "SYNC",
+                        "since": "8.1.0"
```
```suggestion
                        "since": "9.0.0"
```

PingXie commented on tests/unit/cluster/cluster-flush-slot.tcl:
```diff
@@ -0,0 +1,34 @@
+        R 0 CLUSTER FLUSHSLOT $key_slot ASYNC
+        assert_equal [R 0 CLUSTER COUNTKEYSINSLOT $key_slot] 0
+    }
+}
```
Can you add a test case to validate the command is indeed replicated as-is, as opposed to the key deletions getting replicated?

wuranxx replied:
Of course. Added.

PingXie reviewed (commented):
Overall LGTM. We should also add official documentation for this new command.

hwware committed:
Fix bool issue

Signed-off-by: hwware <wen.hui.ware@gmail.com>

hwware committed:
Adjust format for cluster_legacy.c

Signed-off-by: hwware <wen.hui.ware@gmail.com>

hwware committed:
Adjust code format for cluster_legacy.c

Signed-off-by: hwware <wen.hui.ware@gmail.com>

hwware committed:
Remove newline

Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx committed:
Update src/cluster_legacy.c

Co-authored-by: Ping Xie <pingxie@outlook.com>
Signed-off-by: wuranxx <155042264+wuranxx@users.noreply.github.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx committed:
Update src/cluster_legacy.c

Co-authored-by: Ping Xie <pingxie@outlook.com>
Signed-off-by: wuranxx <155042264+wuranxx@users.noreply.github.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx committed:
Update src/commands/cluster-flushslot.json

Co-authored-by: Ping Xie <pingxie@outlook.com>
Signed-off-by: wuranxx <155042264+wuranxx@users.noreply.github.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx committed:
Update src/commands/cluster-flushslot.json

Co-authored-by: Ping Xie <pingxie@outlook.com>
Signed-off-by: wuranxx <155042264+wuranxx@users.noreply.github.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx committed:
Update src/commands/cluster-flushslot.json

Co-authored-by: Ping Xie <pingxie@outlook.com>
Signed-off-by: wuranxx <155042264+wuranxx@users.noreply.github.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx committed:
Update src/commands/cluster-flushslot.json

Co-authored-by: Ping Xie <pingxie@outlook.com>
Signed-off-by: wuranxx <155042264+wuranxx@users.noreply.github.com>
Signed-off-by: hwware <wen.hui.ware@gmail.com>

wuranxx committed:
add flushslot command propagate test.
fix `commands.def`.

Signed-off-by: wuranxx <wuranxx@foxmail.com>

zuiderkwast commented on tests/unit/cluster/cluster-flush-slot.tcl:
```diff
@@ -32,3 +32,33 @@ start_cluster 2 2 {tags {external:skip cluster}} {
+        # not del cmd
+        assert { ![string match "*cmdstat_del*" $info] }
+        # has flushslot cmd
+        assert { [string match "*cmdstat_cluster|flushslot:calls=1*" $info] }
```
Use `assert_no_match` and `assert_match` instead.

wuranxx replied:
fixed.

wuranxx committed:
change tcl test use assert_match and assert_no_match.

Signed-off-by: wuranxx <wuranxx@foxmail.com>

zuiderkwast commented on src/cluster.h:
```diff
@@ -1,3 +1,5 @@
+
 #ifndef __CLUSTER_H
 #define __CLUSTER_H
 
```
`#ifndef __CLUSTER_H` should be the first line in the file, I think.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -6535,13 +6535,14 @@ void removeChannelsInSlot(unsigned int slot) {
 /* Remove all the keys in the specified hash slot.
  * The number of removed items is returned. */
-unsigned int delKeysInSlot(unsigned int hashslot) {
+unsigned int delKeysInSlot(unsigned int hashslot, int lazy, bool is_cmd) {
```
I like the functionality here but I'm not sure about the name of the `is_cmd` argument. It controls two separate things:

is_cmd false: 

* Propagate the deletions using DEL commands
* Don't send keyspace events

is_cmd true:

* Don't propagate deletions to replicas
* Send keyspace events

To send an implicit FLUSHSLOT command from atomic slot migration when a slot migration is aborted, I guess we want to propagate it as a FLUSHSLOT and not send keyspace events. Replicating FLUSHSLOT can be done using a separate call.

Maybe we should have two separate options instead of `is_cmd`.

wuranxx replied:
We should provide two parameters, one is `propagate_del` control  the `del` command is propagated, and the other is `send_del_event` control whether  the `del` keyspace event is sent.

When `cluster flushslot`  is executed, it is confirmed that the `propagate_del` is false, and `send_del_event` should be true?

zuiderkwast replied:
These two parameters sound good to me.

> When `cluster flushslot` is executed, it is confirmed that the `propagate_del` is false, and `send_del_event` should be true?

Yes, I confirm.

The command `cluster flushslot` itself is replicated (separate from this function) so we don't need to replicate individual DEL commands. And yes, send events because the keys are really deleted, as opposed to some scenarios in slot migration where the keys are considered to be moved to another node and not really deleted from the cluster. I read this in a code comment.

zuiderkwast replied:
@murphyjacob4 can you confirm, for ASM when replicating an aborted slot migration, you need to set these parameters to (false, false) right?

wuranxx replied:
added.

zuiderkwast reviewed (commented):
I think this is very close to being merged.

There is a merge conflict that was caused by #1671 that needs to be resolved.

zuiderkwast commented:
I noticed there is unfortunately already a command with a very similar name: [CLUSTER FLUSHSLOTS](https://valkey.io/commands/cluster-flushslots/):

> The CLUSTER FLUSHSLOTS deletes all information about slots from the connected node. It can only be called when the database is empty.

The new command in this PR `CLUSTER FLUSHSLOT` is maybe too similar to that command?

Maybe we should actually make it a top-level command `FLUSHSLOT` without CLUSTER? It will match `FLUSHDB` and `FLUSHALL`.

wuranxx committed:
Merge branch 'unstable' into cluster-flush-slot

madolson commented:
Weekly meeting notes. We will keep the command as is, but we will document this as primarily an internal and debugging command.

PingXie commented:
>Weekly meeting notes. We will keep the command as is, but we will document this as primarily an internal and debugging command.

I think we discussed two decisions

1. whether this should be a top level command?

based on the use case described in the issue, this is clearly not a "data access" command meant for the application developers, hence the guidance of keeping it under `cluster`.

2. whether it is a good idea to introduce a new command whose name differs from that of an existing command by just one character? `cluster flushshot` vs `cluster flushshots`.

I don't remember if we reached conclusion?

zuiderkwast commented:
> I don't remember if we reached a conclusion?

Thanks Ping for summarizing the discussion. We did reach the conclusion to stick to `cluster flushslot`, as Madelyn wrote.

Neither of these commands are frequently used so the similarity between the two commands (`cluster flushshot` vs `cluster flushshots`) is not a major concern.

hwware commented:
> > Weekly meeting notes. We will keep the command as is, but we will document this as primarily an internal and debugging command.
> 
> I think we discussed two decisions
> 
> 1. whether this should be a top level command?
> 
> based on the use case described in the issue, this is clearly not a "data access" command meant for the application developers, hence the guidance of keeping it under `cluster`.
> 
> 2. whether it is a good idea to introduce a new command whose name differs from that of an existing command by just one character? `cluster flushshot` vs `cluster flushshots`.
> 
> I don't remember if we reached a conclusion?

I forgot if you drop off meeting earlier, we have one conclusion that the new command "cluster flushslot" as an internal command.

hwware commented:
> > > Weekly meeting notes. We will keep the command as is, but we will document this as primarily an internal and debugging command.
> > 
> > 
> > I think we discussed two decisions
> > 
> > 1. whether this should be a top level command?
> > 
> > based on the use case described in the issue, this is clearly not a "data access" command meant for the application developers, hence the guidance of keeping it under `cluster`.
> > 
> > 2. whether it is a good idea to introduce a new command whose name differs from that of an existing command by just one character? `cluster flushshot` vs `cluster flushshots`.
> > 
> > I don't remember if we reached a conclusion?
> 
> I forgot if you drop off meeting earlier, we have one conclusion that the new command "cluster flushslot" as an internal command, like cluster-SYNCSLOTS (a new internal command too)

wuranxx committed:
Merge branch 'unstable' into cluster-flush-slot

wuranxx committed:
add `propagate_del` and `send_del_event` parameters to `delKeysInSlot`.

Signed-off-by: wuranxx <wuranxx@foxmail.com>

zuiderkwast commented on src/commands/cluster-flushslot.json:
```diff
@@ -0,0 +1,47 @@
+                        "name": "async",
+                        "type": "pure-token",
+                        "token": "ASYNC",
+                        "since": "9.0.0"
```
Nit: We don't need "since" for the options that are added command itself is new. The command itself has "since": "9.0.0".

wuranxx replied:
fixed.

zuiderkwast reviewed (approved):
LGTM.

@murphyjacob4 Flush one slot at a time, is that enough for cancelled atomic slot migration? If you have 1000 ongoing slot migrations that are cancelled, will you need to send 1000 flushslot commands to the replicas at the same time?

murphyjacob4 commented:
> @murphyjacob4 Flush one slot at a time, is that enough for cancelled atomic slot migration? If you have 1000 ongoing slot migrations that are cancelled, will you need to send 1000 flushslot commands to the replicas at the same time?

I think it is fine. ASM actually has a few scenarios for FLUSHSLOT:

1. If a slot migration is cancelled, each slot within this migration will be removed by the target primary and FLUSHSLOT will be propagated to the target replica(s)
2. If FLUSHSLOT is run on the source primary, it will be sent to the target primary (via slot migration link) foreach ongoing slot export on that node (if any). Target primary will forward this to the target replicas.
3. If FLUSHSLOT is run on the target primary (by user), it will be denied if the slot is being imported (should be covered by this CL I think, will return -MOVED)
4. If FLUSHDB is run on the source primary, if there is a slot export occurring, it will replicate as a FLUSHSLOT command for each owned slot. For all target primaries and their replicas, the FLUSHSLOT commands that match the ongoing slot migration will be propagated via the path in #2
5. If FLUSHDB is run on the target primary, if there is a slot import occurring, it will replicate as a FLUSHSLOT command for each owned slot. Locally, we will leave the importing data. We will not replicate a FLUSHSLOT for the ongoing slot import (since we want to keep that data

In the worst case, it would be something like ~700KiB of replication data to send 16384 FLUSHSLOT commands (each would be <= 43 bytes, `*3\r\n$7\r\nCLUSTER\r\n$9\r\nFLUSHSLOT\r\n$5\r\n16383`). But this seems like a very rare case. We have a path to optimizing with something like `CLUSTER FLUSHSLOTSRANGE` a la `CLUSTER ADDSLOTSRANGE`, but I don't think it is required, and just FLUSHSLOT should better than propagating DEL for each key (at least for 99% of use cases)

zuiderkwast commented:
> * If FLUSHDB is run on the source primary, if there is a slot export occurring, it will replicate as a FLUSHSLOT command for each owned slot. For all target primaries and their replicas, the FLUSHSLOT commands that match the ongoing slot migration will be propagated via the path in 2
> * If FLUSHDB is run on the target primary, if there is a slot import occurring, it will replicate as a FLUSHSLOT command for each owned slot. Locally, we will leave the importing data. We will not replicate a FLUSHSLOT for the ongoing slot import (since we want to keep that data

@murphyjacob4 Now that we have multi-db in cluster mode, FLUSHSLOT will affects all databases. FLUSHDB should only flush the currently SELECT'ed database.

wuranxx committed:
fix code format.

Signed-off-by: wuranxx <wuranxx@foxmail.com>

wuranxx committed:
Merge branch 'valkey-io:unstable' into cluster-flush-slot

wuranxx committed:
Remove `since` from `cluster-flushslot` options definition.

Signed-off-by: wuranxx <wuranxx@foxmail.com>

soloestoy commented:
For the rollback of atomic slot migration, using `FLUSHSLOT` to affect all databases is more appropriate. However, considering the replication of `FLUSHDB` during migration and the possibility that users might need to clear a specific slot in a particular database, I think `CLUSTER FLUSHSLOT <slotid> [DB id]` would be a good option.

madolson commented:
@valkey-io/core-team Shouldn't this have had a major decision tag?

zuiderkwast commented:
The issue has the major decision tag. Seemed enough...(?)

> it doesn't look like it was really restricted to just be internal.

If we don't add a markdown file for it in the docs, it will not show up in the docs, but maybe it will show up somewhere else? How can we make it completely internal?
